x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/slm/SnapshotRetentionConfiguration.java

```diff
@@ -114,32 +116,48 @@ public Integer getMaximumSnapshotCount() {
      */
     public Predicate<SnapshotInfo> getSnapshotDeletionPredicate(final List<SnapshotInfo> allSnapshots) {
         final int snapCount = allSnapshots.size();
-        List<SnapshotInfo> sortedSnapshots = allSnapshots.stream()
+        final List<SnapshotInfo> sortedSnapshots = allSnapshots.stream()
             .sorted(Comparator.comparingLong(SnapshotInfo::startTime))
             .collect(Collectors.toList());
+        final long nonFailedSnapshotCount = allSnapshots.stream()
+            .filter(snap -> SnapshotState.SUCCESS.equals(snap.state()))
+            .count();
+        final long newestSuccessfulTimestamp;
+        if (sortedSnapshots.isEmpty() == false) {
+            newestSuccessfulTimestamp = sortedSnapshots.get(sortedSnapshots.size() - 1).startTime();
```

Comment:
This does not validate that the `sortedSnapshots.get(...)` is actually successful, what if the newest snapshot were `FAILED`? Shouldn't this be something like:

```java
List<SnapshotInfo> reverseSortedSnapshots = new ArrayList<>(sortedSnapshots);
Collections.reverse(reverseSortedSnapshots);
long newestSuccessfulTimestamp = reverseSortedSnapshots.stream()
  .filter(snap -> SnapshotState.SUCCESS.equal(snap.state()))
  .findFirst()
  .orElse(Long.MIN_VALUE);
```

Reply:
Aaaand this is why we have code reviews. Thanks for catching this ❤️